test/AludelV2.ts

```diff
@@ -0,0 +1,3332 @@
+
+const { expect } = chai;
+
+describe("AludelV2", function () {
+  let accounts: SignerWithAddress[], admin: SignerWithAddress;
+  let user: Wallet;
+
+  let powerSwitchFactory: Contract,
+    rewardPoolFactory: Contract,
+    vaultFactory: Contract,
+    stakingToken: Contract,
+    rewardToken: Contract,
+    bonusToken: Contract,
+    aludelFactory: Contract,
+    aludelV2Template: Contract,
+    powered: ContractFactory;
+
+  let template: Contract;
+
+  const mockTokenSupply = ethers.utils.parseEther("1000");
+  const BASE_SHARES_PER_WEI = 1000000;
+  const DAY = 24 * 3600;
+  const YEAR = 365 * DAY;
+  const rewardScaling = { floor: 33, ceiling: 100, time: 60 * DAY };
+
+  let amplInitialSupply: BigNumber;
+
+  const stake = async (
+    user: Wallet,
+    geyser: Contract,
+    vault: Contract,
+    stakingToken: Contract,
+    amount: BigNumberish,
+    vaultNonce?: BigNumberish
+  ) => {
+    // sign permission
+    const signedPermission = await signPermission(
+      "Lock",
+      vault,
+      user,
+      geyser.address,
+      stakingToken.address,
+      amount,
+      vaultNonce
+    );
+    // stake on geyser
+    return geyser.stake(vault.address, amount, signedPermission);
+  };
+
+  const unstakeAndClaim = async (
+    user: Wallet,
+    geyser: Contract,
+    vault: Contract,
+    stakingToken: Contract,
+    amount: BigNumberish,
+    vaultNonce?: BigNumberish
+  ) => {
+    // sign permission
+    const signedPermission = await signPermission(
+      "Unlock",
+      vault,
+      user,
+      geyser.address,
+      stakingToken.address,
+      amount,
+      vaultNonce
+    );
+    // unstake on geyser
+    return geyser.unstakeAndClaim(vault.address, amount, signedPermission);
+  };
+
+  function calculateExpectedReward(
+    stakeAmount: BigNumber,
+    stakeDuration: BigNumberish,
+    rewardAvailable: BigNumber,
+    otherStakeUnits: BigNumberish
+  ) {
+    const stakeUnits = stakeAmount.mul(stakeDuration);
+    const baseReward = rewardAvailable
+      .mul(stakeUnits)
+      .div(stakeUnits.add(otherStakeUnits));
+    const minReward = baseReward.mul(rewardScaling.floor).div(100);
+    const bonusReward = baseReward
+      .mul(rewardScaling.ceiling - rewardScaling.floor)
+      .mul(stakeDuration)
+      .div(rewardScaling.time)
+      .div(100);
+    return stakeDuration >= rewardScaling.time
+      ? baseReward
+      : minReward.add(bonusReward);
+  }
+
+  async function launchProgram(
+    startTime: BigNumberish,
+    _bonusTokens: Contract[],
+    owner: SignerWithAddress,
+    args: any[]
+  ): Promise<Contract> {
+    const deployParams = new AbiCoder().encode(
+      [
+        "address",
+        "address",
+        "address",
+        "address",
+        "uint256",
+        "uint256",
+        "uint256",
+      ],
+      args
+    );
+
+    const factory = AludelFactory__factory.connect(
+      aludelFactory.address,
+      admin
+    );
+    const address = await factory.callStatic.launch(
```

Comment:
why is the callstatic approach comfier than getting the address from the event?

Reply:
I think it's cleaner than things like:
```
      const event = receipt.events?.find(
        (it: Event) => it.address == factory.address && it.event == "ProgramAdded"
      );
      const address = event?.args?.program;
```

Reply:
while it's not something that could really happen in a test, the static call approach is less robust because it's not atomic, and  works by assuming nothing happens between the static call and the regular call, and parsing the event should always work instead.

I don't think it's worth it to modify it, but it was a worthwhile comment